### PR TITLE
[FEATURE] Support columnar shuffle for RSS shuffle writer

### PR DESCRIPTION
**Pull request by kerwin-zk**
### Related issues.
#520 

### How was this patch tested?

/cc @related-reviewer

/assign @main-reviewer

### CODE REVIEW

File: client-spark/shuffle-manager-3/src/main/java/org/apache/spark/shuffle/rss/HashBasedShuffleWriter.java [hunk 1]
```diff
@@ -310,36 +355,65 @@ private void close() throws IOException {
...
-                sendBuffers[i],
-                0,
-                size,
-                numMappers,
-                numPartitions);
-        // free buffer
-        sendBuffers[i] = null;
-        mapStatusLengths[i].add(bytesWritten);
-        writeMetrics.incBytesWritten(bytesWritten);
+    if (isColumnarShuffle) {
```
Comment (waitinfuture) [inline client-spark/shuffle-manager-3/src/main/java/org/apache/spark/shuffle/rss/HashBasedShuffleWriter.java:358]: ditto

**Review by waitinfuture — APPROVED** [review]
LGTM, thanks!

File: client-spark/shuffle-manager-3/src/main/java/org/apache/spark/shuffle/rss/HashBasedShuffleWriter.java [hunk 2]
```diff
@@ -150,12 +160,21 @@ public HashBasedShuffleWriter(
             rssShuffleClient,
             writeMetrics::incBytesWritten,
             mapStatusLengths);
+
+    if (RssConf.columnarShuffleEnabled(rssConf)) {
+      this.schema = SparkUtils.getShuffleDependencySchema(dep);
+      this.isColumnarShuffle = RssColumnarBatchBuilder.supportsColumnarType(schema);
```
Comment (RexXiong) [inline client-spark/shuffle-manager-3/src/main/java/org/apache/spark/shuffle/rss/HashBasedShuffleWriter.java:166]: RssColumnarBatchBuilder Initialization is only required if this condition is true

File: client-spark/shuffle-manager-3/src/main/java/org/apache/spark/shuffle/rss/HashBasedShuffleWriter.java [hunk 3]
```diff
@@ -165,7 +184,11 @@ public void write(scala.collection.Iterator<Product2<K, V>> records) throws IOEx
     } else {
       write0(records);
     }
-    close();
+    if (isColumnarShuffle) {
```
Comment (RexXiong) [inline client-spark/shuffle-manager-3/src/main/java/org/apache/spark/shuffle/rss/HashBasedShuffleWriter.java:187]: I suggest putting closeColumnarWrite in close method，and we could reuse some code at the same time.

**Comment by waitinfuture** [comment]
Plz avoid force push for sake of tracking commits.

**Pull request merged**